### ISSUE TITLE
Unable to get AVS Token

### ISSUE THREAD

**Issue by seattlecajun** [dolanmiu/MMM-awesome-alexa#57]
Hi there, I've setup an account and created an 'app' for the AVS on the Amazon developer's site but I'm not able to complete the setup by getting the access token to generate the config.

I'm able to get to the Amazon login page and see the prompt about 'when you click okay, we'll provide Magic Mirror connectivity to Amazon Alexa" ('Magic Mirror' is the name of my AVS 'app').

Unfortunately, when I hit okay, I'm directed to an Amazon Error page stating "we're sorry! an error has occurred when we tried to process your request, etc".

I've tried the method using the auth_code.sh and auth_code2.sh bash scripts but ultimately received the same 'we're sorry' error message from Amazon.

Any idea where I could have gone wrong?

**Comment by dolanmiu** [comment]
When it says:

```
"we're sorry! an error has occurred when we tried to process your request, etc"
```

Can you expand that message?

Then paste it here

**Comment by seattlecajun** [comment]
Hi @dolanmiu, unfortunately, no. It's not an in-depth error message that offers any actual details - just a generic page with that wording.

![amazon error](https://user-images.githubusercontent.com/17693661/34693970-4cdb2a78-f47a-11e7-8508-97c24be93eb9.PNG)

**Comment by dolanmiu** [comment]
Normally when I get that error, its because the URLs weren't set properly on Amazon's side

**Comment by vivekpk007** [comment]
Same error here, any fixes ??

**Comment by nitrogen73** [comment]
The same issue im also facing here

**Comment by devmil** [comment]
I have exactly the same problem.

**Comment by dolanmiu** [comment]
Getting the refresh token is out of scope of this MMM module

I will re-think of a way to make this process easier

**Comment by devmil** [comment]
Can you hint at any other method of getting the token?

**Comment by dolanmiu** [comment]
The methods are on the README, personally I use the tool I built (mentioned in the README):

https://magic-mirror-avs.github.io/Alexa-Web-Helper/

**Comment by devmil** [comment]
I did the method 2 using ARC (https://github.com/dolanmiu/MMM-awesome-alexa/wiki/How-to-get-AVS-Token-2) and that one was working. No idea what went wrong with the Alexa-Web-Helper

**Comment by dolanmiu** [comment]
So you are all set? Got the refresh token?

**Comment by nitrogen73** [comment]
Whats the problem behind this? Is that any error related to Access token?

transiting to state: idle
transiting to state: listening
transiting to state: busy
{ Error: socket hang up
    at createHangUpError (_http_client.js:252:15)
    at TLSSocket.socketOnEnd (_http_client.js:344:23)
    at emitNone (events.js:91:20)
    at TLSSocket.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:974:12)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9) code: 'ECONNRESET' }

**Comment by devmil** [comment]
Yes, I have a refresh token now. Got it through the second alternative (not using the helper from Awesome Alexa)

**Comment by nitrogen73** [comment]
Im stuck at this point? Can anyone help me please? Im a beginner

transiting to state: idle
transiting to state: listening
transiting to state: busy
{ Error: socket hang up
    at createHangUpError (_http_client.js:252:15)
    at TLSSocket.socketOnEnd (_http_client.js:344:23)
    at emitNone (events.js:91:20)
    at TLSSocket.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:974:12)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9) code: 'ECONNRESET' }

**Comment by DaDonkDon** [comment]
Hello,

just readed all issues to the topic to get the refresh token.
1. Tried  https://magic-mirror-avs.github.io/Alexa-Web-Helper/, but it seems it doesn´t work for me
2. Tried the scripts:

```
CLIENT_ID="amzn1.application-oa2..."
DEVICE_TYPE_ID="pi..."
DEVICE_SERIAL_NUMBER=123
REDIRECT_URI="https://localhost:9745/authresponse"
RESPONSE_TYPE="code"
SCOPE="alexa:all"
SCOPE_DATA="{\"alexa:all\": {\"productID\": \"$DEVICE_TYPE_ID\", \"productInstanceAttributes\": {\"deviceSerialNumber\": \"${DEVICE_SERIAL_NUMBER}\"}}}"

function urlencode() {
  perl -MURI::Escape -ne 'chomp;print uri_escape($_),"\n"'
}

AUTH_URL="https://www.amazon.com/ap/oa?client_id=${CLIENT_ID}&scope=$(echo $SCOPE | urlencode)&scope_data=$(echo $SCOPE_DATA | urlencode)&response_type=${RESPONSE_TYPE}&redirect_uri=$(echo $REDIRECT_URI | urlencode)"
xdg-open ${AUTH_URL}
```



Also the url is filled @amazon: 
![image](https://user-images.githubusercontent.com/35845642/35450534-9f29687c-02c0-11e8-950d-26f1289b8145.png)

First errors:
```
pi@pisander:~/MagicMirror/modules/MMM-awesome-alexa/help/bash-scripts $ chmod aug+x auth_code.sh
pi@pisander:~/MagicMirror/modules/MMM-awesome-alexa/help/bash-scripts $ ./auth_code.sh
Can't locate URI/Escape.pm in @INC (you may need to install the URI::Escape module) (@INC contains: /etc/perl /usr/local/lib/arm-linux-gnueabihf/perl/5.24.1 /usr/local/share/perl/5.24.1 /usr/lib/arm-linux-gnueabihf/perl5/5.24 /usr/share/perl5 /usr/lib/arm-linux-gnueabihf/perl/5.24 /usr/share/perl/5.24 /usr/local/lib/site_perl /usr/lib/arm-linux-gnueabihf/perl-base).
BEGIN failed--compilation aborted.
Can't locate URI/Escape.pm in @INC (you may need to install the URI::Escape module) (@INC contains: /etc/perl /usr/local/lib/arm-linux-gnueabihf/perl/5.24.1 /usr/local/share/perl/5.24.1 /usr/lib/arm-linux-gnueabihf/perl5/5.24 /usr/share/perl5 /usr/lib/arm-linux-gnueabihf/perl/5.24 /usr/share/perl/5.24 /usr/local/lib/site_perl /usr/lib/arm-linux-gnueabihf/perl-base).
BEGIN failed--compilation aborted.
Can't locate URI/Escape.pm in @INC (you may need to install the URI::Escape module) (@INC contains: /etc/perl /usr/local/lib/arm-linux-gnueabihf/perl/5.24.1 /usr/local/share/perl/5.24.1 /usr/lib/arm-linux-gnueabihf/perl5/5.24 /usr/share/perl5 /usr/lib/arm-linux-gnueabihf/perl/5.24 /usr/share/perl/5.24 /usr/local/lib/site_perl /usr/lib/arm-linux-gnueabihf/perl-base).
BEGIN failed--compilation aborted.
[7487:7487:0126/172856.140633:ERROR:sandbox_linux.cc(344)] InitializeSandbox() called with multiple threads in process gpu-process.
[1:12:0126/172856.217032:ERROR:adm_helpers.cc(62)] Failed to query stereo recording.
```

But Chromium opens and shows this:
![image](https://user-images.githubusercontent.com/35845642/35451084-5405dda6-02c2-11e8-8094-6f8992cb794a.png)

Can anybody help me to get a solution? Thanks!

**Comment by Albiorix2016** [comment]
Im having the same issue. I've tried to use the helper site from the pi and i've tried both the other methods as well as the scripts. with the website methods i get the generic amazon error message 

> 

We're sorry!An error occurred when we tried to process your request. Rest assured, we're already working on the problem and expect to resolve it shortly.
--

> 

with the script im getting 

`./auth_code.sh: line 11: syntax error near unexpected token `('
./auth_code.sh: line 11: `  perl -MURI::Escape -ne 'chomp;print uri_escape($),"\n"''
`

i really am not sure where to go from here

**Comment by E3V3A** [comment]
What is this `DEVICE_SERIAL_NUMBER=123` ??
I didn't see any serial number anywhere in the AWS Dashboard...

**Comment by henrikra** [comment]
@E3V3A  Check my node script PR and check its code. It explains it

**Comment by E3V3A** [comment]
@henrikra  Sorry, but your PR doesn't say anything about that. Or it may be that I can't find the right PR, since I can only see one. 

@DaDonkDon  You are missing the PERL module needed to run the script. You should ask the repo owners to document how to install that. (Hint: mpan,cpan,cpanp )

**Comment by henrikra** [comment]
@E3V3A Here is the line you are looking for: https://github.com/dolanmiu/MMM-awesome-alexa/pull/82/files#diff-644654018d7dc381638fba5eddfbbd8cR23

**Comment by E3V3A** [comment]
@henrikra You could just have said so, no? (Would have saved both of us time and effort looking around.)

**Comment by justjim1220** [comment]
Could someone tell me what the Device ID is?
And, where to find it?
![screenshot 3](https://user-images.githubusercontent.com/27590633/37553003-59f4b4b6-298d-11e8-8f9f-cbca5d6a77a1.png)


I've tried everything and not getting anywhere!

Also, the Allowed Return URL: https://sakirtemel.github.io/MMM-alexa/ keeps giving me this:
![screenshot 1](https://user-images.githubusercontent.com/27590633/37552984-24f1aaee-298d-11e8-96cb-19d60678ebcd.png)
whenever I try to add it to my AVS  Security Profile.

I have made all the changes per @henrikra made, but I can't get any further getting the codes I need to connect to AVS because all three of the processes listed to obtain the refesh tokens either ask for a Device ID and/or refer to the URL that gives error at Amazon.

What am I doing wrong or what step(s) am I skipping or missing or possibly overlooking???

**Comment by E3V3A** [comment]
I agree with @justjim1220  and have the same exact problem.

I have tried all 3 of the options (and countless variations) available in the (rather confusing and unclear installation instructions). I have even tried using other tools. Still I am not able to get the `code` needed to get the `refresh_token`. Pasting the result URL in my browser ask me to login and then I get the same error as @seattlecajun s first picture.

The [installation](https://github.com/dolanmiu/MMM-awesome-alexa/blob/master/docs/Installation.md) instructions are not very clear since they are all using different URLs. 

I suggest the maintainers and developers of this repo to please have a look at your [own instructions and follow them](https://github.com/MichMich/MagicMirror/wiki/Documentation-HowTo) from scratch, and come back and help us walk through this and explain how to do this right.  :heart: 

* Is this an URL, return URI or origin problem?
* Updated AWS service problem?
* lacking doc problem?

I'm just about to give up on this project...

**Comment by E3V3A** [comment]
Ok, I've rewritten the Bash script and found out that the reason it doesn't work for so many people is the following. You should read the Amazon documents more carefully and fix the installation docs.

https://developer.amazon.com/docs/alexa-voice-service/authorize-companion-site.html#lwa
https://developer.amazon.com/docs/login-with-amazon/authorization-code-grant.html#Access%20Token%20Request

1. The first script is missing the `state` request parameter:

> state: An opaque value used by your client to maintain state between the request and the response. 
> The authorization service will include this value when redirecting the user back to the client. It is also 
> used to prevent cross-site request forgery. For more information, see Cross-site Request Forgery.

When they say **opaque**, they mean a random 256 bit value, or simply 32 random hex values, but can probably be less. 

2. The request JSON is badly formed with spaces. There should be no spaces in the request string.

3. The `code` token generated:
   * can **only be used once** in the next *refresh_token* script! (If the wget request fails, you need a new code.)
   * probably has a timeout as well

4. **The refresh and access tokens only lasts for 1 hour** (3600 seconds) 

> The access token is valid for one hour. When the access token expires or is about to expire you can exchange the refresh token for new access and refresh tokens.

Thus, I would guess, in order to keep this running, we need both tokens?

5. The **`ProductID` is NOT the `Amazon ID`**, it is the ID you gave to your device or mobile app.

---

@dolanmiu @henrikra 
I hope this helps and that someone can do some fixes. 

PS. Why don't you merge the node script authentication so that it is more easily available/fixable.

**Comment by henrikra** [comment]
@dolanmiu Can you do final testing to the node script PR so we can merge it? :D @E3V3A Is the pr good for you too?

**Comment by E3V3A** [comment]
@henrikra 
> Is the pr good for you too?

TBH, IDK. 
I tried it, and it didn't work, so I went back to use the bash scripts to try to figure out what was wrong. Then I discovered I'd tried to use the Amazon ID instead of ProductID, but by that time I already made a few modifications and improvements (as mentioned above) to the bash scripts. Which then worked. In conclusion, I don't know if they would work for me. 

However, eventually we should try to Node all the scripts, so that it can be automated into the app itself, given that we already give the config info there. It's a bit annoying to have to enter almost the same data in at least 3 different places. I still think you should merge those scripts into the main, but don't remove the bash, yet.

**Comment by henrikra** [comment]
@E3V3A But why did you try to use `Amazon ID` when  the node script asks for `ProductID`?

**Comment by justjim1220** [comment]
Ok, I guess I'm a bit confused at this point...
from @E3V3A 's last post regarding LWS... I'm getting that for anyone to use this module, it has to be a Web App or a device in which I have to set up a User Login.

What I want is for my Family MagicMirror, one used by several people in my family, And I don't want everyone to have to have their own Aamzon Login.

I just want to be able to use Alexa within the mirror as an Interactive Personal Assistant. One that will have Facial recognition, or Voice Recognition, or Both being able to know which family member is currently wanting to use the Mirror. 

I don't have a website to attach to Alexa requiring a User Login. All I need is for Alexa to interact with my family in the same manner as Echo. 

Basically, I would like make my mirror to interact exactly like an Echo device, but not have to purchase an Echo device. 

So, am I in the right place within these forums? 
Can anyone point me in the right direction?
Can anyone give me any ideas of how to incorporate this the way I am wanting it to work?

**Comment by henrikra** [comment]
@justjim1220 Try this branch https://github.com/dolanmiu/MMM-awesome-alexa/pull/82/files

**Comment by E3V3A** [comment]
@justjim1220 
>  I would like make my mirror to interact exactly like an Echo device

That's what we all want. But it take some time getting used to the various jargon, interfaces and developer sites (Amazon). 

It works pretty close to Echo, but after you create the device and have it connected to your MM, then you need to login to the Alexa setting site. This is different from the developer site. There you can train the voice, setup new skills and location etc. 

https://alexa.amazon.com/spa/index.html#cards

**Comment by iamshadmirza** [comment]
I'm trying to generate refresh token. When I am copy pasting the link to the browser to authorize Alexa , I'm getting a blank webpage with message:
``{"error":"MissingParams","message":"The following parameters were missing or empty strings: state"}``
I tried Alexa helper module for token generation, still getting same error.

**Comment by onkararj** [comment]
im getting 


"This site can’t be reached

localhost refused to connect.

this error
help plz

**Comment by dolanmiu** [comment]
This error is intended

All the info you need is in the address bar
